### ISSUE TITLE
bug: GC race condition: agent/future worker threads not registered as mutators, pending AgentMsg closures not traced

### ISSUE THREAD

**Issue by csm** [csm/clojurust#89]
## Summary

The tracing GC has a memory-safety bug that manifests when GC fires while agent or future worker threads are executing Clojure code. The bug has two independent parts that together allow the GC to free objects still in use by worker threads.

This is currently latent in production: the default soft limit (75% of ¼ RAM, ~3 GiB on a 16 GiB machine) is high enough that most workloads don't trigger it. However any long-running program that uses agents or futures under allocation pressure will eventually hit it. In the AOT test harness we worked around it by keeping allocation volume below the threshold (interleaved require→run-tests→remove-ns keeps live objects ~300 MiB), but that is not a general fix.

Discovered while fixing the Clojure test suite CI (stack overflow + OOM) in PR #92 / branch `claude/async-plan-phase-f-Gg4Xi`.

---

## Part 1 — Worker threads are not registered as GC mutators

Agent worker threads are spawned with `std::thread::spawn` in `handle_agent` (`cljrs-interp/src/apply.rs:892`). They never call `register_mutator()`, so `GcCancellation::registered_threads` is 0 for them.

When the test/main thread initiates STW collection (`begin_stw`), `wait_for_threads_to_park` computes:

```rust
let expected = CONFIG_CANCELLATION.registered_threads().saturating_sub(1);
// expected = 0  ← no registered threads
```

It returns immediately without waiting. GC starts while the worker thread may be mid-execution, holding live `GcPtr` values in its local Rust stack variables and `ALLOC_ROOTS` that the collector never sees. The collector can free objects the worker thread is actively using → use-after-free → UB.

The same issue applies to future worker threads.

**Affected code:**
- `cljrs-interp/src/apply.rs` — `handle_agent` (line ~892), future spawning
- `cljrs-gc/src/cancellation.rs` — `wait_for_threads_to_park`, `safepoint`

---

## Part 2 — Pending `AgentMsg` closures are not traceable

Even if Part 1 is fixed and worker threads park at safepoints, `AgentFn` closures sitting in the MPSC channel buffer between `send` and `recv` capture live `GcPtr` values:

```rust
// handle_send, cljrs-interp/src/apply.rs ~line 936
let agent_fn: AgentFn = Box::new(move |state| {
    // f, extra, agent_val_clone all captured — GcPtr values inside
    apply_value(&f, call_args, &mut call_env)
    ...
});
channel.send(AgentMsg::Update(agent_fn));
```

`Agent::trace` (`cljrs-value/src/types.rs:993`) traces `state`, `error`, and `watches` but **not** `sender`. Standard Rust channels don't expose their internal buffer for iteration, so the captured `GcPtr` values (`f: Value`, `extra: Vec<Value>`, `agent_val_clone`) are invisible to the GC. Once the calling thread's `pop_alloc_frame` clears those values from its own `ALLOC_ROOTS`, they are unreachable from any traced root — the GC can free them while the worker thread is about to execute that action.

**Affected code:**
- `cljrs-value/src/types.rs` — `Agent` struct, `Trace for Agent`
- `cljrs-interp/src/apply.rs` — `handle_send`

---

### Fix 1 — Thread registration + cross-thread ALLOC\_ROOTS sharing

Agent and future threads must call `register_mutator()` so GC waits for them to park. Additionally, `safepoint()` must donate the parking thread's `ALLOC_ROOTS` to a shared structure before parking, so the collector can trace in-flight allocations from all threads — not just the one that initiated collection:

```rust
// cancellation.rs
pub fn safepoint() {
    if !CONFIG_CANCELLATION.in_progress() { return; }
    // Expose this thread's in-flight GcPtr roots before parking.
    cljrs_gc::donate_alloc_roots_for_gc();
    CONFIG_CANCELLATION.park();
    while CONFIG_CANCELLATION.in_progress() { std::thread::yield_now(); }
    CONFIG_CANCELLATION.unpark();
}
```

This requires:
- A `Mutex<Vec<*mut GcBoxHeader>>` global structure for donated roots
- `cljrs_gc::donate_alloc_roots_for_gc()` copies the calling thread's `ALLOC_ROOTS` into it
- The collector traces from this structure in addition to its own roots
- Structure is cleared after each collection

### Fix 2 — Make pending agent actions GC-visible

Replace the opaque `SyncSender<AgentFn>` with a bare signal channel plus a GC-visible pending queue on the `Agent` struct:

```rust
pub struct Agent {
    pub state: Arc<Mutex<Value>>,
    pub error: Arc<Mutex<Option<Value>>>,
    /// Signal channel — worker blocks until a token arrives, then pops from `pending`.
    pub sender: Mutex<SyncSender<()>>,
    /// GC-visible queue of pending (function, extra-args) pairs.
    pub pending: Arc<Mutex<VecDeque<(Value, Vec<Value>)>>>,
    pub watches: Mutex<Vec<(Value, Value)>>,
}

impl Trace for Agent {
    fn trace(&self, visitor: &mut MarkVisitor) {
        self.state.lock().unwrap().trace(visitor);
        if let Some(e) = self.error.lock().unwrap().as_ref() { e.trace(visitor); }
        let pending = self.pending.lock().unwrap();
        for (f, args) in pending.iter() {
            f.trace(visitor);
            for arg in args { arg.trace(visitor); }
        }
        for (k, f) in self.watches.lock().unwrap().iter() {
            k.trace(visitor); f.trace(visitor);
        }
    }
}
```

`handle_send` pushes `(f, extra)` to `agent.pending` and sends `()` to the signal channel. The worker thread receives the signal, pops from `pending`, and executes. Because `pending` is traced by `Agent::trace`, and the agent is always reachable from caller roots while in use, the captured values are never invisible to the GC.

---

## Files to change

| File | Change |
|---|---|
| `crates/cljrs-gc/src/lib.rs` | Add `donate_alloc_roots_for_gc()` API and global donated-roots structure; trace donated roots in `collect()` |
| `crates/cljrs-gc/src/cancellation.rs` | Call `donate_alloc_roots_for_gc()` in `safepoint()` before parking |
| `crates/cljrs-value/src/types.rs` | Restructure `Agent` with `pending: VecDeque<(Value, Vec<Value>)>`; update `Trace` impl |
| `crates/cljrs-interp/src/apply.rs` | `handle_agent`: call `register_mutator()` in worker thread body; `handle_send`: push to `pending` queue instead of capturing in closure |
| Wherever futures are spawned | Same `register_mutator()` treatment |

---

## Reproducer (approximate)

```clojure
;; Requires enough allocation pressure to cross the GC soft limit
;; (~3 GiB tracked on a 16 GiB machine with default config).
;; With a lower ESTIMATED_OBJECT_SIZE or reduced soft limit this
;; manifests much sooner.

(let [result (promise)
      a (agent 0)]
  (send a (fn [s]
            (deliver result :done)
            (inc s)))
  ;; ... many allocations here to trigger GC ...
  @result)  ; may hang if GC freed the captured fn before worker ran it
```

**Comment by csm** [comment]
This bug is possibly obsolete given the new focus on isolates in async workers.
